### PR TITLE
[Comb] Merge ors of concats with constant operands.

### PR DESCRIPTION
**Pull request by richardxia**
Fixes one of the patterns described in #1624, where a common pattern is the or'ing of two values that only have bits set in non-overlapping positions.

This is the first time I've contributed to this codebase, and it's been a long time since I've written C++, so I'm happy to make any changes that are requested.

The approach I took was to simultaneously iterate over the operands of both concats in a left-to-right fashion, resolving the result of the `or`, but it got a bit tricky due to having to split up the constant 0's into narrower constants when the operand bit ranges don't align between the two concats. Please let me know if there's a better or simpler approach to this, since my `while` loop ended up with many different edge cases and it doesn't seem the most elegant.

I also realized that we can handle constant 1's in the concats at the same time by having them mask over any variable bits, but I wasn't sure if that should be done in a different pass or handled at the same time.

Lastly, I'm still coming up to speed on the different classes and methods available, so apologies in advance if I'm using the wrong ones or if I've reimplemented something that's already available as a helper method somewhere.

### CODE REVIEW

File: lib/Dialect/Comb/CombFolds.cpp
Comment (darthscsi) [inline lib/Dialect/Comb/CombFolds.cpp:945]: nit: llvm coding style preferes prefix increment.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (darthscsi) [inline lib/Dialect/Comb/CombFolds.cpp:942]: createOrFold might uniq constants.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (darthscsi) [inline lib/Dialect/Comb/CombFolds.cpp:922]: don't need to make  copy of the operands just to iterate.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (darthscsi) [inline lib/Dialect/Comb/CombFolds.cpp:899]: How aggressive should this be? only the tail?  all pairs? something in between?  Since you check early for overlap, going over all pairs might be fine.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (darthscsi) [inline lib/Dialect/Comb/CombFolds.cpp:929]: you could just getDefiningOp and not store operand[1,2].  the constantop's type would work below, as would adding it to the new operands.

**Review by darthscsi — COMMENTED** [review]
Thanks!

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp:945]: Sorry about that, habit that was ingrained into me many years ago when first learning how to program. Will fix in a sec.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp:922]: Actually, I had to make a copy in order for my `*it1 = ...` assignments to work in the loop iterations when I truncate the constants. It probably isn't very kosher of me to copy the entire array just so that I can use it as temporary storage for the truncated values, but I couldn't think of a cleaner way to deal with it.

Would it better for me to just declare a variable outside the loop to hold the current operand? Something like:

```c++
auto it1 = operands1.begin();
auto operand1 = *it1;
++it1;

while (it1 != end1 && it2 != end2) {
  ...

  // Case where I need to truncate an operand
  operand1 = rewriter.create<hw::ConstantOp>(..., cst1.getValue().trunc(...));

  // Case where I want to advance the iterator;
  ++it1;
  operand1 = *it1;

  ...
}
```

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp:899]: Sounds good, I can check all pairs, and if that ends up being too slow in some cases, we can put some limit on it.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp:929]: I actually end up using `operand[1,2]` in the case where one of the current operands is a constant 0 and the other is some other value, since I directly push the other, non-ConstantOp value to `newConcatOperands`.

I think what I'd really love here is something like pattern matching in a functional programming language where I extract out just the constant value in the case where it's a constant and where I only have the opaque type in the cases where it's not a ConstantOp.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp]: This has fixed in my latest, amended commit.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp]: Thanks, all of these ConstantOp create's have been updated to use createOrFold.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp]: My latest commit now iterates over all pairs of `ConcatOp`s.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp:924]: I tried doing this in https://github.com/richardxia/circt/commit/e0d9f74e14fddd4d6130bf783ddbb255c7c70848, but I have not pushed up to this branch since it ended up being very messy with having to constantly check whether it was safe to dereference the iterator pointer.

Is there a better way for me to iterate over the original operands while preserving the ability to maintain a bit of local state for the current amount of truncation of the constants?

**Comment by richardxia** [comment]
Thanks for the feedback! I amended and repushed my commit to this branch addressing some of your comments, but I'm a bit stuck on the others. If there's anything I can do to save you time reviewing my code, like studying up on a specific part of the codebase, let me know.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (lattner) [inline lib/Dialect/Comb/CombFolds.cpp:1044]: nice

File: lib/Dialect/Comb/CombFolds.cpp
Comment (lattner) [inline lib/Dialect/Comb/CombFolds.cpp:894]: Plz add an example in the comment, it will make it easier to know at-a-glance what sort of thing this is doing.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (lattner) [inline lib/Dialect/Comb/CombFolds.cpp:911]: I'd recommend taking a slightly different approach on this, which would make this a bit simpler and more general, building on the fact that `rewriter.createOrFold` autofolds and that canonicalize itself is iterative.

Instead of using computeKnownBits, just scan both concats to see if either operand has any constants as operands to the concat.  If so, turn:

`or (concat ..., concat ...)`

into: `concat(or ..., or ..., or ..., or ...)` where the chunks of the result are the maximum contiguous ranges from the two concats you have in.  When folding the nested or's, extracts from constants will automatically fold into smaller constants, and any "or" of 0 or 111 will fold automatically.

Does this sort of approach make sense to you?

**Review by lattner — COMMENTED** [review]
Your code is nice and clean Richard, I'm impressed!

File: lib/Dialect/Comb/CombFolds.cpp
Comment (lattner) [reply lib/Dialect/Comb/CombFolds.cpp:911]: by max contiguous chunks, what I mean is that something like `concat(a_3bit, b_2bit)|concat(cst_2bit, d_3bit)` would turn into:

```
concat(or(extract(a_3bit, 2..1), cst_2bit),
              or(extract(a_3bit, 0...0), extract(d_3bit, 2..2)),
              or(b_2bit, extract(d_3bit, 1..0)))
```

Each of those things would nicely fold themselves.

Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp:911]: > where the chunks of the result are the maximum contiguous ranges from the two concats you have in.

Just to double check my understanding, are you saying we do this in the following example?

```
or(concat(x: i2, 0: i6), concat(0: i5, y: i3))
->
concat(or(x: i2, 0: i2), or(0: i3, 0: i3), or(0: i3, y: i3))
```

And in the case where x and y overlap, then we should extract pieces of the unknown values? For example:

```
or(concat(x: i6, 0: i2), concat(0: i1, y: i7))
->
// Using Verilog slice notation
concat(or(x[5:5], 0): i1, or(x[4:0], y[6:2]): i5, or(0, y[1:0]): i2)
```

If so, yeah, that makes sense to me, and I like that it allows the constant folding of the individual `or`s to happen automatically. I can work on making these changes.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp:1044]: I can't take credit for this, since I saw this pattern being used for the `canonicalizeLogicalCstWithConcat()` function just a few lines above this, but I agree that it's pretty neat.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp:894]: Yep, will do.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp:911]: Somehow my browser tab only showed your first comment but not the second when I wrote up that comment. Sounds like I'm understanding it correctly, so let me work on that.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp]: Added a pair of examples to this in my latest push.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp]: Thanks, this worked out very well, and I've pushed up an amended commit that takes this approach.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp]: This has been addressed in my latest commit, where I used @lattner's approach of extracting subranges of the operands, where I no longer need to mutate the operands since I can now treat ConstantOps and other ops uniformly.

**Comment by richardxia** [comment]
Thanks again for all the feedback! I've significantly changed the code to use the approach of extracting subranges of operands, which is a lot cleaner and more general. I've renamed the pull request and the commit message, since it's no longer just about `or`'s of exclusive `concat` ops.

I also added one more test case to demonstrate that this also works with constant `1`s as well.

I tried checking this against the second pattern that @lattner mentioned in https://github.com/llvm/circt/issues/1624, the one with the mux, but it looks like it's still unable to optimize it yet because of the mux. I think if we could canonicalize `mux(cond, 2'h2, 2'h0)` into `concat(mux(cond, 1'h1 : 1'h0), 1'h0)`, then the changes in this PR would be able to simplify that for free. Is that something I should create a separate ticket for?

Other than that, I think this is ready for another review.

**Comment by darthscsi** [comment]
> I think if we could canonicalize `mux(cond, 2'h2, 2'h0)` into `concat(mux(cond, 1'h1 : 1'h0), 1'h0)`, 

This kind of mux narrowing seems a good thing.  A separate PR for it is fine (and these are both small and related, so folding it into this one would be fine too, but this looks self contained and done, so I'd do a separate PR).

File: test/Dialect/Comb/canonicalization.mlir
Comment (darthscsi) [inline test/Dialect/Comb/canonicalization.mlir:104]: Can you add a case for an or > 2 operands?

**Review by darthscsi — APPROVED** [review]
Nice!

File: test/Dialect/Comb/canonicalization.mlir
Comment (richardxia) [reply test/Dialect/Comb/canonicalization.mlir:104]: Yep, no problem just added a test case with 3 operands.

I also took the opportunity to rebase this on top of the latest `main`, since several bumps to llvm submodule have happened in the meantime.

**Comment by richardxia** [comment]
By the way, I haven't requested write access yet (was planning to after landing this patch), so I'll need someone else to actually hit the merge button. Thanks again for looking at this PR!

File: test/Dialect/Comb/canonicalization.mlir
Comment (darthscsi) [reply test/Dialect/Comb/canonicalization.mlir:104]: Rebasing isn't necessary if you don't have conflicts.  Squash merge is used exclusively so there is a linear history with 1 PR == 1 commit.

Comment (richardxia) [reply test/Dialect/Comb/canonicalization.mlir:104]: Got it. I'm never sure what GitHub squash merges will do with merges in the feature branch, so I've always erred on the side of rebasing my branches when keeping up to date with the main branch, but good to know.

**Comment by lattner** [comment]
It's no problem to request commit access, plz just read the llvm developer policy and follow the instructions.

File: lib/Dialect/Comb/CombFolds.cpp
Comment (lattner) [inline lib/Dialect/Comb/CombFolds.cpp:929]: Random drive-by, but I'd move the short-circuit up.  If hasConstantOp1 is true, then no need to compute hasConstantOp2.  Something like:

if (!hasConstantOp1) {
  if (!anyof(...))
    return false;
}

would look nice.

**Review by lattner — COMMENTED** [review]
This is great work Richard, thank you for tackling this!

File: lib/Dialect/Comb/CombFolds.cpp
Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp:929]: Sounds good, I can quickly make this change, and will keep this kind of thing in mind in the future.

Comment (richardxia) [reply lib/Dialect/Comb/CombFolds.cpp:929]: I ended up committing this directly to main in 00a3c0d723605e4a932f55187007bd98cd94a30b, which I'm assuming was OK for this kind of change, based on what I'm reading in the [LLVM Developer Policy](https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access), but let me know if I should submit pull requests for this kind of thing in the future.

Comment (lattner) [reply lib/Dialect/Comb/CombFolds.cpp:929]: Thanks!  LGTM

**Pull request merged**
